main/ControllerTask.cpp

```diff
@@ -23,99 +24,174 @@
-		const float MAX_REFER_CHANGE  = 50.0f;
-		const float BEZIER_SMOOTHNESS =  0.1f;
+		// loop variables
+		float err;
+		float error_integral = 0.0f;
+		float error_derivative;
+		float u;
+		float exec_time_us;
 
 		while (true) {
-			(void)xQueueReceive(motor_speed_q, &motor_speed, 10);
-			(void)xQueueReceive(refer_speed_q, &refer_speed, 10);
-
-			prev_motor_speeds[motor_idx] = motor_speed;
-			motor_idx = (motor_idx+1)%N_PREV_SPEEDS;
-			motor_speed = 
-				std::accumulate(prev_motor_speeds, prev_motor_speeds+N_PREV_SPEEDS, 0.0f)
-				/N_PREV_SPEEDS;
-
-			if (std::abs(refer_speed - bezier_speed_ref) > rpm2rad_s(MAX_REFER_CHANGE)) {
-				float P0 = bezier_speed_ref;
-				float P1 = bezier_speed_ref + BEZIER_SMOOTHNESS;
-				float P2 = refer_speed      - BEZIER_SMOOTHNESS;
-				float P3 = refer_speed;
-
-				bezier_speed_ref = bezierCurve(t, P0, P1, P2, P3);
-				t += transition_step;
+			(void)xQueueReceive(motor_speed_q, &motor_speed, QUEUE_TIMEOUT);
+			(void)xQueueReceive(refer_speed_q, &refer_speed, QUEUE_TIMEOUT);
+
+			task_st = std::chrono::high_resolution_clock::now();
+			motor_speed = motorFilter(motor_speed);
+			refer_speed = refFilter(refer_speed);
+
+			err = refer_speed - motor_speed;
+
+			if (motor_speed < rpm2rad_s(50)) {
+				u = OUT_MIN;
 			}
 			else {
-				t = 0.0f;
-				bezier_speed_ref = refer_speed;
+				error_integral = (OUT_MIN < u && u < OUT_MAX) ? ierror(err) : error_integral;
+				u = mamdani(rad_s2rpm(err)) + K_I*error_integral;
+				u = uFilter(u);
 			}
+			error_derivative = derror(err);
 
-			float err = bezier_speed_ref - motor_speed;
-			float u   = pid(err);
-			float real_u = u;
-
-			const float U_MIN = 0.17f, U_MAX = 0.95f;
-			uint8_t pwm_out = (uint8_t)(std::clamp(u, U_MIN, U_MAX)*PWM_MAX);
+			uint8_t pwm_out = (uint8_t)(std::clamp((u-U_MIN)/(U_MAX-U_MIN), OUT_MIN, OUT_MAX)*PWM_MAX);
 			pwm_out &= PWM_MAX;
 
-			pwm_set_duty(LEDC_CHANNEL_0, pwm_out);
-
-			char buffer[29+21+1] = {0};
-			int  offset = 0;
-			offset  = sprintf(buffer       ,"\rR:%6.2f ", rad_s2rpm(refer_speed));
-			offset += sprintf(buffer+offset,  "M:%6.2f ", rad_s2rpm(motor_speed));
-			offset += sprintf(buffer+offset,  "e:%9.2f ", err);
-			offset += sprintf(buffer+offset,  "=> u:%9.2e o: %2d", real_u, pwm_out);
-			(void)printf("%s", buffer);
-
-			vTaskDelay(80 / portTICK_PERIOD_MS);
+			pwm_set_duty(pwm_channel, pwm_out);
+
+			task_en = std::chrono::high_resolution_clock::now();
+			microseconds task_duration_us = duration_cast<microseconds>(task_en-task_st);
+			exec_time_us = task_duration_us.count();
+
+			(void)xQueueOverwrite(_tel_ref_speed_q,   &refer_speed);
+			(void)xQueueOverwrite(_tel_motor_speed_q, &motor_speed);
+			(void)xQueueOverwrite(_tel_error_q,       &error_integral);
+			(void)xQueueOverwrite(_tel_error_der_q,   &error_derivative);
+			(void)xQueueOverwrite(_tel_control_signal_q, &u);
+			(void)xQueueOverwrite(_tel_exec_time_q, &exec_time_us);
+
+			// constexpr int BUFF_SIZE = 12+29+10+24+1;
```

Comment:
Instead of just comment it. Create a debug boolean class attribute that can be passed in the constructor to enable debug or not.  If you dont wanna do by constructor you can play also with #ifdefs